engine/resources/pippet_run.go

```diff
@@ -0,0 +1,160 @@
+// Mgmt
+// Copyright (C) 2013-2018+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os/exec"
+	"sync"
+)
+
+// PippetRunner is the interface to the wrapped "puppet yamlresource receive"
+// process that keeps running alongside mgmt. Individual resources are expected
+// to register themselves, and can use the Apply function from that point on.
+type PippetRunner interface {
+	Register(interface{}) error
+	Unregister(interface{}) error
+	Apply(string, string, string) PippetResult
+	Close() error
+}
+
+// PippetResult is the structured return value type for the PippetRunner's
+// Apply function.
+type PippetResult struct {
+	Error     bool
+	Failed    bool
+	Changed   bool
+	Exception string
+}
+
+type pippetReceiver struct {
+	stdin         io.WriteCloser
+	stdout        io.ReadCloser
+	registerMutex sync.Mutex
+	applyMutex    sync.Mutex
+	registered    int
+}
+
+var pippetRunnerInstance PippetRunner
```

Comment:
global vars (if required) should be at the top of the file, right after imports. Basically so they're easily seen, since they can be problematic if you aren't expecting them.

Reply:
Fixed.